### PR TITLE
[FLINK-5423] Implement Stochastic Outlier Selection

### PR DESCRIPTION
**Pull request by Fokko**
Implemented the Stochastic Outlier Selection algorithm in the Machine Learning library, including the test code. Added documentation.

Thanks for contributing to Apache Flink. Before you open your pull request, please take the following check list into consideration.
If your changes take all of the items into account, feel free to open your pull request. For more information and/or questions please refer to the [How To Contribute guide](http://flink.apache.org/how-to-contribute.html).
In addition to going through the list, please provide a meaningful description of your changes.

- [x] General
  - The pull request references the related JIRA issue ("[FLINK-XXX] Jira title text")
  - The pull request addresses only one issue
  - Each commit in the PR has a meaningful commit message (including the JIRA id)

- [x] Documentation
  - Documentation has been added for new functionality
  - Old documentation affected by the pull request has been updated
  - JavaDoc for public methods has been added

- [x] Tests & Build
  - Functionality added by the pull request is covered by tests
  - `mvn clean verify` has been executed successfully locally or a Travis build has passed

### CODE REVIEW

**Comment by Fokko** [comment]
I think we have some flakey tests, since it passes on my own travis:
https://travis-ci.org/Fokko/flink/builds/189855914

**Comment by tillrohrmann** [comment]
Thanks for your contribution @Fokko. I'll take a look at this PR in the next days :-)

File: flink-libraries/flink-ml/src/test/scala/org/apache/flink/ml/outlier/StochasticOutlierSelectionITSuite.scala [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+      BreezeLabeledVector(1, BreezeDenseVector(Array(5.0, 1.0)))
+    ))
+
+    val distanceMatrix = StochasticOutlierSelection
+      .computeDissimilarityVectors(data)
+      .map(_.data)
+      .collect()
+      .toArray
+
+    print(distanceMatrix)
```
Comment (tillrohrmann) [inline flink-libraries/flink-ml/src/test/scala/org/apache/flink/ml/outlier/StochasticOutlierSelectionITSuite.scala]: I think we should remove the print statement

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/outlier/StochasticOutlierSelection.scala [hunk 2]
```diff
@@ -0,0 +1,367 @@
...
+  = {
+    new TransformDataSetOperation[StochasticOutlierSelection, T, Double] {
+      override def transformDataSet(instance: StochasticOutlierSelection,
+                                    transformParameters: ParameterMap,
+                                    input: DataSet[T]): DataSet[Double] = {
+
+        val resultingParameters = instance.parameters ++ transformParameters
+
+        // Map to the right format
+        val vectorsWithIndex = input.zipWithIndex.map(vector => {
```
Comment (tillrohrmann) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/outlier/StochasticOutlierSelection.scala]: Does it strictly have to be the index you're zipping with? Would be a unique ID be enough? If so, then we could use the `DatatSetUtils.zipWithUniqueId` method which is considerably cheaper than `zipWithIndex`.

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/outlier/StochasticOutlierSelection.scala [hunk 3]
```diff
@@ -0,0 +1,367 @@
...
+
+  def apply(): StochasticOutlierSelection = {
+    new StochasticOutlierSelection()
+  }
+
+  // ===================================== Operations ==============================================
+  case class BreezeLabeledVector(idx: Int, data: BreezeVector[Double])
+
+  implicit val transformLabeledVectors = {
+    new TransformDataSetOperation[StochasticOutlierSelection, LabeledVector, (Int, Double)] {
```
Comment (tillrohrmann) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/outlier/StochasticOutlierSelection.scala]: Maybe we could add a comment that the label of `LabeledVector` should be the index or a unique integer value.

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/outlier/StochasticOutlierSelection.scala [hunk 4]
```diff
@@ -0,0 +1,367 @@
...
+    * @return Returns the affinity vector of the input vector.
+    */
+  def binarySearch(dissimilarityVector: BreezeVector[Double],
+                   logPerplexity: Double,
+                   maxIterations: Int,
+                   tolerance: Double,
+                   beta: Double = 1.0,
+                   betaMin: Double = Double.NegativeInfinity,
+                   betaMax: Double = Double.PositiveInfinity,
+                   iteration: Int = 0): BreezeVector[Double] = {
```
Comment (tillrohrmann) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/outlier/StochasticOutlierSelection.scala]: The non written down coding style was to break a long parameter list by writing each parameter in its own line with double indentation. Something like
```
def fun(
    x: Int,
    y: Double)
  : String = {
  // foobar
}

**Comment by Fokko** [comment]
Thanks @tillrohrmann, excellent idea regarding the documentation. I'll also process the code comments, good feedback. Somewhere today or tomorrow I will fix this.

Cheers, Fokko

File: flink-libraries/flink-ml/src/test/scala/org/apache/flink/ml/outlier/StochasticOutlierSelectionITSuite.scala (hunk 1 above)
Comment (Fokko) [inline flink-libraries/flink-ml/src/test/scala/org/apache/flink/ml/outlier/StochasticOutlierSelectionITSuite.scala]: Oops, still in there from the debugging.

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/outlier/StochasticOutlierSelection.scala (hunk 3 above)
Comment (Fokko) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/outlier/StochasticOutlierSelection.scala]: Good thinking, I've added a scaladoc.

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/outlier/StochasticOutlierSelection.scala (hunk 2 above)
Comment (Fokko) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/outlier/StochasticOutlierSelection.scala]: Yes, I've checked this and it is possible.

File: flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/outlier/StochasticOutlierSelection.scala (hunk 4 above)
Comment (Fokko) [inline flink-libraries/flink-ml/src/main/scala/org/apache/flink/ml/outlier/StochasticOutlierSelection.scala]: Check

**Comment by Fokko** [comment]
@tillrohrmann I've added documentation about the algorithm. Can you check? 

Cheers, Fokko

**Comment by tillrohrmann** [comment]
Changes look really good to me :-) Travis is passing modulo an unrelated test failure. Will merge this PR. Thanks a lot for your contribution @Fokko.

**Pull request closed without merging**